android-components/components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt

```diff
@@ -145,6 +145,22 @@ interface AwesomeBar {
      * It returns a list of [Suggestion]s to be displayed by the [AwesomeBar].
      */
     interface SuggestionProvider {
+        /**
+         * Represents priority order for [SuggestionProvider]s.
+         **/
+        enum class Type {
+            SEARCH_ENGINE,
+            SEARCH_ACTION,
+            SEARCH_SUGGESTION,
+            COMBINED_HISTORY,
+            HISTORY,
+            SYNCED_TAB,
+            SESSION,
+            BOOKMARKS,
+            SHORTCUTS,
+            CLIPBOARD,
+        }
```

Comment:
I am open to further discussion here. I had imagined that this `enum` with the int priority would exist in the consumer (Fenix). By defining these `Type`s within the interface, we are basically saying these are the only available suggestion types, however, I don't think that would be our intention since this is just a concept that could be implemented by the consumer depending on whatever they want to do.

Reply:
@kycn I would love to further validate these thoughts if you have any input as well.

Reply:
Agree that this enum (which isn't extendable) shouldn't stay in `concept`.
An interface should not know who implements it and as you've said clients are free to come with their own implementations on top of what the interface knows at this time.
If the intent is to force developers think about choosing the right priority in relation to others - and not use just any `Int` this approach helps though. Using a wrapper over the `Int` goes half way to solve developer's haste.

I am not a fan of another `enum` since it introduces a new type for developers to reason about and choose which could lead to miss matching.
I see myself having to read quite a bit on the providers to know how to choose the right priority to use between `SEARCH_ENGINE`, `SEARCH_ACTION`, `SEARCH_SUGGESTION`.
For clearer mappings, since these priorities would be used when a new provider instance is constructed we could use a simple map of each provider class and it's priority and avoid having to reason about the new enum types.
In Fenix a top level const:
```kotlin
val awesomeBarSuggestionsPriorities: Map<KClass<out AwesomeBar.SuggestionProvider>, Int> = mapOf(
    SearchActionProvider::class to 1,
    SearchEngineSuggestionProvider::class to 2
    ....
)
```
(with the provider class's default being `0`)

from which we can easily read and use the priority with 
```
SearchActionProvider(
    store = components.core.store,
    searchUseCase = searchUseCase,
    icon = searchBitmap,
    showDescription = false,
    priority = awesomeBarSuggestionsPriorities[SearchActionProvider::class],
)
```

(As an alternative to the above map we could also declare in Fenix extension `priority` properties for each provider type. This would also ensure no involuntary miss matching.)

Reply:
Hey @gabrielluong thanks for notifying! After reading [your comments in Fenix side](https://github.com/mozilla-mobile/fenix/pull/28024#pullrequestreview-1201297534), I strongly agree that the current approach is not as flexible as it should be. The maintenance is hard and as you mentioned is error-prone for future additions. In the simplest use case, where we would want to introduce a new provider (with a new type), this approach would require us to publish a new a-c release with the new type. This should be avoided for the reasons you described in your comment I've just referenced. The concept of a "concept" in a-c is conflicting with the idea of providing a list of order for suggestion providers. That is another reason why enum should be avoided here The concept modules are public apis that in general should behave quite generic and should not provide any implementation details. 
Addressing to the flexibility you mentioned in Fenix pr, I unfortunately do not have a very strong opinion that solves all the "what if" questions you asked reasonably. 
When I was going over the code I realized maybe we can introduce a new ProviderSorter interface/api in `AwesomeBar` feature module and implement it on Fenix end (could be in `AwesomeBarWrapper)` by injecting any sort of sorting mechanism that we like. This sorter then maybe issued to Awesomebar (in a-c) over the `addProvider` and `removeProvider` apis via additional (null-default) parameters. This is not my final suggestion, i would need to think more about this as it definitely needs more investigation/refinement and discussion I believe.

Reply:
I am not against having a `ProviderSorter` interface. I should also point out some of my suggested use cases are hypothetical at the moment (eg, the requirement to have a dynamic sorting depending on which suggestions are shown). I am thankful that we're considering potential use cases because flexibility is always nice.

However, I don't want us to block the current issues at hand for too long. So, I am looking to suggest a simpler path forward in the meantime and we can file a followup about introducing a more throughout solution. If we went back to the original commit https://github.com/mozilla-mobile/firefox-android/commit/b2053b2b6e029110085ec31a302caf966bd42fae that introduces the `priority` as Int, remove the hardcoded `priority`, add a breaking change note in the changelog, and added the following enum in Fenix, and use the `ProviderPriority` to set the `priority` when the suggestion provider classes are initialized in Fenix.

```
enum class ProviderPriority(val priority: Int) {
    BOOKMARKS(5),
    COMBINED_HISTORY(4),
    HISTORY_STORAGE(4),
    SEARCH_ACTION(1),
    SEARCH_ENGINE(0),
    SEARCH(3),
    SYNC_TABS(6)
}
```

How does that sound?

Reply:
Sounds good to me! My only question would be, then, what about when all the providers get null values as priority? Or will we make the priority fileds of the providers non-null? 
We were sorting the providers according to their titles previously, could it be a default fallback in case we use nullable priority fields for providers and the consumer of a-c would not want to provide any sort of sorting logic with the integers?

Reply:
Just wanted to mention out here that @csadilek and @rocketsroger suggested that it would be better if we could sort the groups using already defined suggestion scores (as mentioned in my above comment) for most of the providers. Will be moving forward with this and see if we could have the results this way .